packages/hash/shared/src/save.ts

```diff
@@ -1,693 +1,473 @@
-
-    const updatedEntities = entities.filter(
-      (_, position) =>
-        !removedBlockEntities.some(
-          ([, removedPosition]) => removedPosition === position,
-        ),
-    );
+const capitalizeComponentName = (cId: string) => {
+  let componentId = cId;
 
-    const actions = removedBlockEntities.map(
-      ([, position], idx): UpdatePageAction => ({
-        /**
-         * Each removal results in the position of further removals being
-         * subtracted by 1 – luckily we can just used the index in the array to
-         * work this out
-         */
-        removeBlock: { position: position - idx },
-      }),
-    );
+  // If there's a trailing slash, remove it
+  const indexLastSlash = componentId.lastIndexOf("/");
+  if (indexLastSlash === componentId.length - 1) {
+    componentId = componentId.slice(0, -1);
+  }
 
-    return [actions, updatedEntities] as const;
-  },
-);
+  //                      *
+  // "https://example.org/value"
+  const indexAfterLastSlash = componentId.lastIndexOf("/") + 1;
+  return (
+    //                      * and uppercase it
+    // "https://example.org/value"
+    componentId.charAt(indexAfterLastSlash).toUpperCase() +
+    //                       ****
+    // "https://example.org/value"
+    componentId.substring(indexAfterLastSlash + 1)
+  );
+};
+
+const randomPlaceholder = () => `placeholder-${uuid()}`;
+
+const flipMap = <K, V>(draftToPlaceholder: Map<K, V>): Map<V, K> =>
+  new Map(
+    Array.from(draftToPlaceholder, ([key, value]) => [value, key] as const),
+  );
+
+type EntityTypeForComponentResult = [string, UpdatePageAction[]];
+
+const hasComponentId = (
+  properties: unknown,
+): properties is { componentId: string } =>
+  typeof properties === "object" &&
+  properties !== null &&
+  "componentId" in properties &&
+  typeof (properties as any).componentId === "string";
+
+const ensureEntityTypeForComponent = async (
+  componentId: string,
+  newTypeAccountId: string,
+  entityTypes: EntityType[],
+  cache: Map<string, string>,
+): Promise<EntityTypeForComponentResult> => {
+  const actions: UpdatePageAction[] = [];
+
+  let desiredEntityTypeId: string | undefined = cache.get(componentId);
+
+  if (!desiredEntityTypeId) {
+    desiredEntityTypeId = entityTypes.find(
+      (type) =>
+        hasComponentId(type.properties) &&
+        type.properties.componentId === componentId,
+    )?.entityId;
+  }
 
-const moveBlocks = defineOperation(
-  (entities: BlockEntity[], draftBlockEntityIds: DraftEntity["entityId"][]) => {
-    const otherExistingBlockEntityIds = draftBlockEntityIds.filter(
-      (blockEntityId) => !!blockEntityId,
+  if (!desiredEntityTypeId) {
+    const componentMeta = await fetchBlockMeta(componentId);
+    const jsonSchema = JSON.parse(
+      JSON.stringify(componentMeta.componentSchema),
     );
 
-    const actions: UpdatePageAction[] = [];
-    entities = [...entities];
+    delete jsonSchema.properties.editableRef;
```

Comment:
This is just moving code, but my understanding is that the schema is generated from react component props, and this can therefore end up in the schema for react apps. Going to leave this as its just moving code.

Reply:
Okay, thank you for explaining - that's good to know. Ideally I think this is something the blocks should be addressing (ensuring `editableRef` doesn't unintentionally end up in their block schema) rather than our backend having to special case any JSON-Schema with a `editableRef` property, which is not a sustainable approach.

Why is it a problem for `editableRef` to be in the JSON-Schema?

If this line is preventing some bug, then I suppose we can leave it for now.

But if it does not, I think we should get rid of this special casing here now, and create an asana task for refactoring the blocks that are currently inadvertently injecting an `editableRef` property into their JSON-Schemas.

I think that would involve them explicitly declaring the block schemas, rather than having them be generated from their react props.